### ISSUE TITLE
Create `MSRunSample` model class

### ISSUE THREAD

**Issue by lparsons** [Princeton-LSI-ResearchComputing/tracebase#708]
### Inspiration

Part of the [TraceBase MS Data Storage Model](https://docs.google.com/document/d/1CUg1fChG3ZEHgMcLZnzq5eRDxiixhhH5quELHXjLIRc/edit?usp=sharing)

### Description

Create a model class `MSRunSample` with the following fields:

* `id` PK int
* `ms_run_id` int FK >- `MSRunSequence.id`
* `sample_id` int FK >- `Sample.id`
* `ms_data_files` int FK >-< `ArchiveFile.id` NULL

![image](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/645128/39e50c7e-34a4-45f4-9268-7f04aac2c9c9)

`PeakGroup` does not need to be updated to connect to this model for this issue, that can be done separately.

### Dependencies

This issue cannot be started until the completion of the following issue(s):

- #693 
- #707

### Comment

-----

### Assumptions

- List of assumptions that the code will not explicitly address/check
- E.g. We will assume input is correct (explaining why there is no validation)

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### Limitations

- A list of things this work will specifically not do
- E.g. This feature will only handle the most frequent use case X

### Affected Components

A tentative list of anticipated repository items that will be changed, labeled
with "add", "delete", or "change".  One item per line.  (Mostly, this will be a
list of files.)

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

#### Interface Change description

Describe changes to usage. *E.g. GUI/command-line changes*

#### Code Change Description

Describe code changes planned for the feature. *(Pseudocode encouraged)*

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

**Comment by hepcat72** [comment]
I was looking over this issue to inform my work on #707, or assess the possibility of doing both issues, one after the other or concurrently.  I have some questions/need some clarity.

I'm having some trouble resolving (in my mind) the data relationships related to the `ms_data_files` field as they relate to:

- peak group records (and peak annotation files)
- mzXML files
- sample data headers in the peak annotation files
- MSRunSequence records

In other words, what represents a unique `MSRunSample`?  I assume the link: `ms_data_files` is for mzXML files?  So if I wanted all of the mzxml files associated with a particular peak annotation file, or a particular peak group, or a particular sequence, I'm having trouble figuring out how that would be structured.  But maybe those are not all associations that we need to support?

Let me highlight what I mean.  So for any one sequence, it is linked with a number of MSRunSample records, and thereby to a number of mzXML files that were a part of that sequence.  And for any one Peak Group, it is linked to a single MSRunSample record, and a single mzXML file.  If a biological sample is represented multiple times in a sequence and only once in a PeakGroup, is the M:M link for every mzXML file for connecting to all mzXML files for that sample in the sequence, or is it for the one mzXML file used for the creation of the peakgroup?

Maybe I'm off, but I kind of feel like an MSRunSample record should be unique to an mzXML file (i.e. an MSRunSample record should link to a single mzXML file record in ArchiveFile) to clear up the relationship from Sequence->MSRunSample->mzXML and PeakGroup->MSRunSample->mzXML.  This would mean that the combination of the 3 IDs would represent a unique record.  Otherwise, in order to support paths from sequence to mzxmls and peakgroup to mzxmls, there wouldn't be a combination of fields to uniquely represent a record.

Also, I could be wrong, but I think that `polarity` data (in the description, not in the image) doesn't (or won't, after current PRs are merged) exist anymore.  Didn't a migration delete the protocol records that contained the polarity?  It's not mentioned in the migration in #709, but given the difference between the description and the image, perhaps it fell through the cracks?  We can always add it in the load.

**Comment by hepcat72** [comment]
In today's Tracebase planning meeting, Michael and I discussed the above.

First, Michael recalled that polarity was removed, so I think that the issue description is just stale.

The second issue is more up in the air for me... (**TL;DR** at the bottom)

Michael reminded me that we had deemed that associating a specific mzXML file with a column of data is not 100% reliable, and that that is likely the reason for the association of a peakgroup to multiple mzxml files instead of 1 in the model depiction above - because we can't guaranty the correct association.

I pointed out though that in any one of our study loads, I noted that the headers have always been unique across the files in the study.  If one sample is represented in multiple peak annotation files, the headers are all different and directly correspond to the names of the mzXML files.  e.g. `sample1.mzxml`, `sample1_pos.mzxml`, and `sample1_neg.mzxml`.

Michael said that in that situation, that is true.  The mapping of sample to header is reliable, but that that may not always be the case.

To start from the beginning..., the instrument enforces a unique sample name for each vial, but if a sample vial is drawn from multiple times (to be run in different modes: neg/pos), the file names are the same.  What distinguishes them is that they are in different directories.  Users append (e.g.) `_pos` to the raw file when they compile files for an analysis.  And when the mzXML files are generated, their names are automatically based on the raw file name, i.e. they automatically inherit those "extensions".  It's also notable that if the instrument ever appends anything to a name (like if they restart a run and the instrument encounters an existing file(/dir?)), it appends a datestamp.

While this is all informative, I'm still unclear as to why we can't associate a specific mzXML file used to create a peakgroup.  If the header and file name correspond (which is a reliable association), and the user provides the mzXML files used in generating the analysis data being loaded, everything should be good.

Based on my understanding, I posit that the issue with distinguishing files is at the point of compiling the RAW files that are used to extract the data into mzXML files.  You can't do it by file name alone.  You have to use the directory.  But whether or not the user modifies the RAW file names, they know what files it is that they are feeding into El-Maven and accucor/isocorr.  When they collect the files, they appear to always append (e.g.) `_pos` to tell them apart, based originally on the directory name, which is useful, but it's not necessary.  It doesn't matter that those file names may match other RAW files that were not used in the analysis.  It only matters that of the files they collected, they either made the names (study-)unique or they made them unique for 1 particular EL-Maven/Accucor analysis and kept track of what files went into each accucor/isocorr file.  All we need, when a user submits a batch of files that went into the analysis, to be able to associate a peakgroup with an mzXML file is the header to mzXML file match, because we're storing the file ourselves.  We don't need to be able to tell them apart from same-named files from the original run.  What matters is that the users are supplying the files that go into the analysis and that the headers generated are based on the file names.  If the user submits the wrong files (e.g. an mzXML file that wasn't a part of the generation of an accucor file, but with a matching name), there's nothing we can do about that - and it would be wrong regardless of what we do.

Could a user feed 2 files with the same name into El-Maven and accucor/isocorr?  Maybe, maybe not.  If they did, they would have trouble knowing which column is which (if they don't encounter an error).  Could a user compile a study that includes 2 accucor files that have same-named columns that come from 2 different but same-named sets of mzXML files?  I'm sure they could, but if that's the case, I don't think it's unreasonable to expect them to keep track of what files went into each analysis.  Nor do I think it's unreasonable to expect that their column names in a single accucor file be unique.  Both of those problems should be solved by the user.  The only thing we need to account for is the possibility that multiple accucor files could have the same headers and same-named mzxml files that actually differ - and we can work around that issue by requiring that they be broken up into separate submissions.  In our experience with these studies, that situation has never come up, though we have seen multiple instances with those appended strings that distinguish all of the mzXML files.  Even if they didn't make them all study-level-unique, the user should know which files went into which El-Maven/accucor analysis.  The only other way around re-used names is to bite the bullet on potential duplicate file uploads and associate mzxml files with each accucor file.

**TL;DR** My point is, the mzXML file name ambiguity only exists *before* El-Maven/Accucor/Isocorr files are generated.  Even if we end up with multiple files with the same name in one study, the user should be expected to know which files went into which analysis and that the names in a particular accucor/isocorr analysis are unique.  They can get it wrong even if we implement the model as described above, so we might as well make the discrete association and not worry about it.

**Comment by lparsons** [comment]
> In other words, what represents a unique MSRunSample? I assume the link: ms_data_files is for mzXML files? So if I wanted all of the mzxml files associated with a particular peak annotation file, or a particular peak group, or a particular sequence, I'm having trouble figuring out how that would be structured. But maybe those are not all associations that we need to support?

An `MSRunSample` is intended to represent the injection of a sample during an `MSRunSequence`. To retrieve all of the `mzXML` files associated with:
- peak group
    ```
    peak_group.msrun_sample.ms_data_files
    ```
- peak annotation file
    ```
    models.ArchiveFile.objects.filter(
        msrunsample__in=models.MSRunSample.objects.filter(
        peakgroup__in=paf2.peakgroup_set.all()
        )
    ).distinct()
    ```
- sequence
    ```
    models.ArchiveFile.objects.filter(
        msrunsample__in=sequence.msrunsample_set.all()
    )
    ```

> If a biological sample is represented multiple times in a sequence and only once in a PeakGroup, is the M:M link for every mzXML file for connecting to all mzXML files for that sample in the sequence, or is it for the one mzXML file used for the creation of the peakgroup?

If a biological sample is represented multiple times in a sequence, that would mean multiple injections of the same sample and thus, generate multiple `MSRunSample` records. It's been my understanding that each injection (`MSRunSample`) generates a single `RAW` file (which we do not store) and 1 or more `mzXML` files (which we would store). The data in the `RAW` files is split based on polairty and "scan settings" to generate these additional files. All of the `mzXML` files for a given injection (one RAW file) would be associated with the `MSRunSample` record for that injection.

I appreciate the additional work to associate `mzXML` files with specific `PeakGroup` records, but I'd rather not rehash this discussion. Let's stick with what we have for now and perhaps keep some notes, etc. on how we could change things going forward if it seems like researchers would benefit from the additional detail. I'd rather get a "good enough" product working soon than delay for the "more perfect" solution.

> Also, I could be wrong, but I think that polarity data (in the description, not in the image) doesn't (or won't, after current PRs are merged) exist anymore.

Yes, you are correct, the polarity should have been removed from the description. Sorry about that, I'll fix that now.

**Comment by hepcat72** [comment]
Merged as of #772 and #773
